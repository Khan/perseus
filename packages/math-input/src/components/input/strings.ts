import * as i18n from "@khanacademy/wonder-blocks-i18n";

import type {MathQuillAriaStatisStringsMap} from "./mathquill-types";

export const mathQuillStrings = {
    mathInputBox: i18n._("Math input box"),
    ariaStaticStringsMap: {
        before: (obj) => i18n._("before %(obj)s", {obj}),
        after: (obj) => i18n._("after %(obj})s", {obj}),
        "beginning of": (obj) => i18n._("beginning of %(obj)s", {obj}),
        "end of": (obj) => i18n._("end of %(obj)s", {obj}),
        Baseline: i18n._("Baseline"),
        Superscript: i18n._("Superscript"),
        selected: (obj) => i18n._("%(obj)s selected", {obj}),
        "no answer": i18n._("no answer"),
        "nothing selected": i18n._("nothing selected"),
        "nothing to the right": i18n._("nothing to the right"),
        "nothing to the left": i18n._("nothing to the left"),
        "block is empty": i18n._("block is empty"),
        "nothing above": i18n._("nothing above"),
        labelValue: (label, value) =>
            i18n._("%(label)s: %(value)s", {label, value}),
    } satisfies MathQuillAriaStatisStringsMap,
};
