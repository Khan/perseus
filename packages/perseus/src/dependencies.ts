import * as React from "react";

import type {PerseusDependencies, PerseusDependenciesV2} from "./types";

let _dependencies: PerseusDependencies | null | undefined = null;

export const setDependencies = (dependencies: PerseusDependencies) => {
    _dependencies = dependencies;
};

export const getDependencies = (): PerseusDependencies => {
    if (_dependencies) {
        return _dependencies;
    }

    throw new Error(
        [
            "Perseus has not been provided required dependencies.",
            "setDependencies(dependencies) must be called first.",
            "Make sure Perseus is being imported from javascript/perseus/perseus.js.",
        ].join("\n"),
    );
};

export type DependencyProps = Partial<
    React.ContextType<typeof DependenciesContext>
>;

export const DependenciesContext = React.createContext<PerseusDependenciesV2>({
    analytics: {onAnalyticsEvent: async () => {}},
});

/**
 * useDependencies provides easy access to the current Perseus dependencies.
 */
export const useDependencies = (): PerseusDependenciesV2 => {
    const deps = React.useContext(DependenciesContext);

    return deps;
};
