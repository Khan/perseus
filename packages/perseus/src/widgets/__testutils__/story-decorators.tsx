import * as React from "react";

export const mobileDecorator = (Story) => (
    <div className="framework-perseus perseus-mobile">
        <Story />
    </div>
);

export const articleDecorator = (Story) => (
    <div className="framework-perseus perseus-article">
        <Story />
    </div>
);

export const mobileArticleDecorator = (Story) => (
    <div className="framework-perseus perseus-mobile perseus-article">
        <Story />
    </div>
);

export const rtlDecorator = (Story) => (
    <div style={{direction: "rtl"}}>
        <Story />
    </div>
);
