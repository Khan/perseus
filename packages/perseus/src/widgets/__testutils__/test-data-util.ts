import type {PerseusItem, PerseusRenderer} from "@khanacademy/perseus-core";

/**
 * Creates a Perseus item with the provided question/PerseusRenderer data.
 * Useful for converting test data used with RendererWithDebugUI to test data
 * for ServerItemRendererWithDebugUI.
 *
 * @param {PerseusRenderer} question - The question to be included in the Perseus item.
 * @returns {PerseusItem} The created Perseus item.
 */
export const createPerseusItem = (question: PerseusRenderer): PerseusItem => {
    return {
        question: question,
        answer: null,
        itemDataVersion: {
            major: 0,
            minor: 1,
        },
        hints: [],
        answerArea: {
            calculator: false,
            chi2Table: false,
            financialCalculatorMonthlyPayment: false,
            financialCalculatorTotalAmount: false,
            financialCalculatorTimeToPayOff: false,
            periodicTable: false,
            periodicTableWithKey: false,
            tTable: false,
            zTable: false,
        },
    };
};
