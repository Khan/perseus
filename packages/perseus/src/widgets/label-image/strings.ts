import * as i18n from "@khanacademy/wonder-blocks-i18n";

export const strings = {
    hideAnswersToggleLabel: i18n._("Hide answer choices"),
};
