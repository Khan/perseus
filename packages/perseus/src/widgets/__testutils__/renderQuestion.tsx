import {RenderStateRoot} from "@khanacademy/wonder-blocks-core";
import {render} from "@testing-library/react";
import * as React from "react";

import {
    testDependenciesV2,
    testDependencies,
    // eslint-disable-next-line import/no-relative-packages
} from "../../../../../testing/test-dependencies";
import {
    DependenciesContext,
    useDependencies,
    setDependencies,
} from "../../dependencies";
import * as Perseus from "../../index";
import {mockStrings} from "../../strings";
import {registerAllWidgetsForTesting} from "../../util/register-all-widgets-for-testing";

import type {APIOptions} from "../../types";
import type {PerseusRenderer} from "@khanacademy/perseus-core";
import type {PropsFor} from "@khanacademy/wonder-blocks-core";

type RenderResult = ReturnType<typeof render>;

type ExtraProps = Omit<PropsFor<typeof Perseus.Renderer>, "strings">;

export const renderQuestion = (
    question: PerseusRenderer,
    apiOptions: APIOptions = Object.freeze({}),
    extraProps?: ExtraProps,
): {
    container: HTMLElement;
    renderer: Perseus.Renderer;
    rerender: (question: PerseusRenderer, extraProps?: ExtraProps) => void;
    unmount: RenderResult["unmount"];
} => {
    setDependencies(testDependencies);
    registerAllWidgetsForTesting();

    let renderer: Perseus.Renderer | null = null;
    const {container, rerender, unmount} = render(
        <RenderStateRoot>
            <DependenciesContext.Provider value={testDependenciesV2}>
                <Renderer
                    ref={(node) => (renderer = node)}
                    question={question as any}
                    apiOptions={apiOptions}
                    extraProps={{
                        ...extraProps,
                        strings: mockStrings,
                    }}
                />
            </DependenciesContext.Provider>
        </RenderStateRoot>,
    );
    // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
    if (!renderer) {
        throw new Error(`Failed to render!`);
    }
    const renderAgain = (
        question: PerseusRenderer,
        extraProps?: ExtraProps,
    ) => {
        rerender(
            <RenderStateRoot>
                <DependenciesContext.Provider value={testDependenciesV2}>
                    <Renderer
                        ref={(node) => (renderer = node)}
                        question={question}
                        apiOptions={apiOptions}
                        extraProps={{
                            ...extraProps,
                            strings: mockStrings,
                        }}
                    />
                </DependenciesContext.Provider>
            </RenderStateRoot>,
        );
        if (!renderer) {
            throw new Error(`Failed to rerender!`);
        }
    };

    return {container, renderer, rerender: renderAgain, unmount};
};

const Renderer = React.forwardRef<
    Perseus.Renderer,
    {
        question: PerseusRenderer;
        apiOptions: APIOptions;
        extraProps?: PropsFor<typeof Perseus.Renderer>;
    }
>((props, ref) => {
    const dependencies = useDependencies();
    return (
        <Perseus.Renderer
            ref={ref}
            content={props.question.content}
            images={props.question.images}
            widgets={props.question.widgets}
            problemNum={0}
            apiOptions={props.apiOptions}
            strings={mockStrings}
            {...props.extraProps}
            {...dependencies}
        />
    );
});
