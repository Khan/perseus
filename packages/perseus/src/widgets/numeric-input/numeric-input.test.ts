import {act, screen} from "@testing-library/react";
import {userEvent as userEventLib} from "@testing-library/user-event";

import {testDependencies} from "../../../../../testing/test-dependencies";
import * as Dependencies from "../../dependencies";
import {mockStrings} from "../../strings";
import {renderQuestion} from "../__testutils__/renderQuestion";

import {
    maybeParsePercentInput,
    NumericInput,
    unionAnswerForms,
} from "./numeric-input";
import {
    question1AndAnswer,
    multipleAnswers,
    percentageProblem,
    multipleAnswersWithDecimals,
    question1,
    duplicatedAnswers,
    withCoefficient,
} from "./numeric-input.testdata";

import type {Rubric} from "./numeric-input";
import type {UserEvent} from "@testing-library/user-event";

describe("numeric-input widget", () => {
    const [question, correct, incorrect] = question1AndAnswer;

    let userEvent: UserEvent;
    beforeEach(() => {
        userEvent = userEventLib.setup({
            advanceTimers: jest.advanceTimersByTime,
        });

        jest.spyOn(Dependencies, "getDependencies").mockReturnValue(
            testDependencies,
        );
    });

    it("Should accept the right answer", async () => {
        // Arrange
        const {renderer} = renderQuestion(question);

        // Act
        await userEvent.type(
            screen.getByRole("textbox", {hidden: true}),
            correct,
        );

        // Assert
        expect(renderer).toHaveBeenAnsweredCorrectly();
    });

    it("Should render predictably", async () => {
        // Arrange
        const {container} = renderQuestion(question);
        expect(container).toMatchSnapshot("first render");

        // Act
        await userEvent.type(
            screen.getByRole("textbox", {hidden: true}),
            correct,
        );

        // Assert
        expect(container).toMatchSnapshot("after interaction");
    });

    it("should reject an incorrect answer", async () => {
        // Arrange
        const {renderer} = renderQuestion(question);

        // Act
        await userEvent.type(
            screen.getByRole("textbox", {hidden: true}),
            incorrect,
        );

        // Assert
        expect(renderer).toHaveBeenAnsweredIncorrectly();
    });
});

describe("static function getOneCorrectAnswerFromRubric", () => {
    beforeEach(() => {
        jest.spyOn(Dependencies, "getDependencies").mockReturnValue(
            testDependencies,
        );
    });

    it("can get one correct answer from a rubric with multiple answers", () => {
        const widget = multipleAnswersWithDecimals.widgets["numeric-input 1"];
        const widgetOptions = widget && widget.options;
        const answers: ReadonlyArray<any> =
            (widgetOptions && widgetOptions.answers) || [];

        const singleAnswer = NumericInput.getOneCorrectAnswerFromRubric({
            answers,
            labelText: "",
            size: "medium",
            static: false,
            coefficient: false,
        });
        expect(singleAnswer).toBe("12.2");
    });

    it("can get one correct answer from a rubric with one answer", () => {
        const widget = question1.widgets["numeric-input 1"];
        const widgetOptions = widget && widget.options;
        const answers: ReadonlyArray<any> =
            (widgetOptions && widgetOptions.answers) || [];
        const singleAnswer = NumericInput.getOneCorrectAnswerFromRubric({
            answers,
            labelText: "",
            size: "medium",
            static: false,
            coefficient: false,
        });
        expect(singleAnswer).toBe("1252");
    });

    it("can not get a correct answer from a rubric with no answer", () => {
        const answers: Array<never> = [];
        const singleAnswer = NumericInput.getOneCorrectAnswerFromRubric({
            answers,
            labelText: "",
            size: "medium",
            static: false,
            coefficient: false,
        });
        expect(singleAnswer).toBeUndefined();
    });

    it("supports error bars", () => {
        const rubric: Rubric = {
            answers: [
                {
                    status: "correct",
                    value: 1.0,
                    maxError: 0.2,
                    answerForms: ["decimal"],
                    simplify: "",
                    strict: false,
                    message: "",
                },
            ],
            labelText: "",
            size: "normal",
            static: false,
            coefficient: true,
        };
        const singleAnswer = NumericInput.getOneCorrectAnswerFromRubric(rubric);
        expect(singleAnswer).toBe("1 Â± 0.2");
    });
});

describe("static function validate", () => {
    beforeEach(() => {
        jest.spyOn(Dependencies, "getDependencies").mockReturnValue(
            testDependencies,
        );
    });

    it("with a simple value", () => {
        const rubric: Rubric = {
            answers: [
                {
                    value: 1,
                    status: "correct",
                    maxError: 0,
                    simplify: "",
                    strict: false,
                    message: "",
                },
            ],
            labelText: "",
            size: "normal",
            static: false,
            coefficient: true,
        };

        const useInput = {
            currentValue: "1",
        } as const;

        const score = NumericInput.validate(useInput, rubric, mockStrings);

        expect(score).toMatchInlineSnapshot(`
            {
              "earned": 1,
              "message": null,
              "total": 1,
              "type": "points",
            }
        `);
    });

    it("with nonsense", () => {
        const rubric: Rubric = {
            answers: [
                {
                    value: 1,
                    status: "correct",
                    maxError: 0,
                    simplify: "",
                    strict: false,
                    message: "",
                },
            ],
            labelText: "",
            size: "normal",
            static: false,
            coefficient: true,
        };

        const useInput = {
            currentValue: "sadasdfas",
        } as const;

        const score = NumericInput.validate(useInput, rubric, mockStrings);

        expect(score).toMatchInlineSnapshot(`
            {
              "message": "We could not understand your answer. Please check your answer for extra text or symbols.",
              "type": "invalid",
            }
        `);
    });

    // Don't default to validating the answer as a pi answer
    // if answerForm isn't set on the answer.
    // The answer value, userInput.currentValue, and
    // the omission of answerForms in the answer are
    // important to the test.
    // https://khanacademy.atlassian.net/browse/LC-691
    it("doesn't default to validating pi", () => {
        const rubric: Rubric = {
            answers: [
                {
                    maxError: null,
                    message: "",
                    simplify: "required",
                    status: "correct",
                    strict: true,
                    value: 45.289,
                },
            ],
            labelText: "",
            size: "normal",
            static: false,
            coefficient: false,
        };

        const userInput = {
            // (pi / 12) * 173 = 45.291
            // within the 0.01 margin of error
            // to trigger the pi validation flow
            currentValue: "45.282",
        } as const;

        const score = NumericInput.validate(userInput, rubric, mockStrings);

        expect(score.message).not.toBe(
            "Your answer is close, but you may " +
                "have approximated pi. Enter your " +
                "answer as a multiple of pi, like " +
                "<code>12\\ \\text{pi}</code> or " +
                "<code>2/3\\ \\text{pi}</code>",
        );
        expect(score.message?.includes("pi")).toBeFalsy();
    });

    it("still validates against pi if provided in answerForms", () => {
        const rubric: Rubric = {
            answers: [
                {
                    maxError: null,
                    message: "",
                    simplify: "required",
                    status: "correct",
                    strict: false,
                    value: 311.01767270538954,
                    answerForms: ["pi"],
                },
            ],
            labelText: "",
            size: "normal",
            static: false,
            coefficient: false,
        };

        const userInput = {
            currentValue: "99 pi",
        } as const;

        const score = NumericInput.validate(userInput, rubric, mockStrings);

        expect(score).toMatchInlineSnapshot(`
        {
          "earned": 1,
          "message": null,
          "total": 1,
          "type": "points",
        }
    `);
    });

    it("with a strict answer", () => {
        const rubric: Rubric = {
            answers: [
                {
                    value: 1,
                    status: "correct",
                    maxError: 0,
                    simplify: "",
                    strict: true,
                    message: "",
                },
            ],
            labelText: "",
            size: "normal",
            static: false,
            coefficient: true,
        };

        const useInput = {
            currentValue: "1.0",
        } as const;

        const score = NumericInput.validate(useInput, rubric, mockStrings);

        expect(score).toMatchInlineSnapshot(`
            {
              "earned": 1,
              "message": null,
              "total": 1,
              "type": "points",
            }
        `);
    });

    it("with a strict answer and max error is outside range", () => {
        const rubric: Rubric = {
            answers: [
                {
                    value: 1,
                    status: "correct",
                    maxError: 0.2,
                    simplify: "",
                    strict: true,
                    message: "",
                },
            ],
            labelText: "",
            size: "normal",
            static: false,
            coefficient: true,
        };

        const useInput = {
            currentValue: "1.3",
        } as const;

        const score = NumericInput.validate(useInput, rubric, mockStrings);

        expect(score).toMatchInlineSnapshot(`
            {
              "earned": 0,
              "message": null,
              "total": 1,
              "type": "points",
            }
        `);
    });

    it("with a strict answer and max error is inside range", () => {
        const rubric: Rubric = {
            answers: [
                {
                    value: 1,
                    status: "correct",
                    maxError: 0.2,
                    simplify: "",
                    strict: true,
                    message: "",
                },
            ],
            labelText: "",
            size: "normal",
            static: false,
            coefficient: true,
        };

        const useInput = {
            currentValue: "1.12",
        } as const;

        const score = NumericInput.validate(useInput, rubric, mockStrings);

        expect(score).toMatchInlineSnapshot(`
            {
              "earned": 1,
              "message": null,
              "total": 1,
              "type": "points",
            }
        `);
    });
});

describe("Numeric input widget", () => {
    let userEvent: UserEvent;
    beforeEach(() => {
        userEvent = userEventLib.setup({
            advanceTimers: jest.advanceTimersByTime,
        });

        jest.spyOn(Dependencies, "getDependencies").mockReturnValue(
            testDependencies,
        );
    });

    it("can handle multiple correct answers (Part one)", async () => {
        // Arrange
        const {renderer} = renderQuestion(multipleAnswers);

        // Act
        await userEvent.type(screen.getByRole("textbox", {hidden: true}), "1");

        // Assert
        expect(renderer).toHaveBeenAnsweredCorrectly();
    });

    it("can handle multiple correct answers (Part two)", async () => {
        // Arrange
        const {renderer} = renderQuestion(multipleAnswers);

        // Act
        await userEvent.type(screen.getByRole("textbox", {hidden: true}), "2");

        // Assert
        expect(renderer).toHaveBeenAnsweredCorrectly();
    });

    it("can handle duplicated answers", async () => {
        // Arrange
        const {renderer} = renderQuestion(duplicatedAnswers);

        // Act
        await userEvent.type(
            screen.getByRole("textbox", {hidden: true}),
            "2.4",
        );

        // Assert
        expect(renderer).toHaveBeenAnsweredCorrectly();
    });

    it("can handle coefficients", async () => {
        // Arrange
        const {renderer} = renderQuestion(withCoefficient);

        // Act
        await userEvent.type(
            screen.getByRole("textbox", {hidden: true}),
            "1.0",
        );

        // Assert
        expect(renderer).toHaveBeenAnsweredCorrectly();
    });

    it("handles answers that are percentages", async () => {
        // Arrange
        const {renderer} = renderQuestion(percentageProblem);

        // Act
        await userEvent.type(
            screen.getByRole("textbox", {hidden: true}),
            "33%",
        );

        // Assert
        expect(renderer).toHaveBeenAnsweredCorrectly();
    });

    it("handles answers that are decimals", async () => {
        // Arrange
        const {renderer} = renderQuestion(multipleAnswersWithDecimals);

        // Act
        await userEvent.type(
            screen.getByRole("textbox", {hidden: true}),
            "12.2",
        );

        // Assert
        expect(renderer).toHaveBeenAnsweredCorrectly();
    });

    it("styles differently on mobile", () => {
        const {container} = renderQuestion(multipleAnswersWithDecimals, {
            // I wish this was more clear but this is how mobile
            // rendering is triggered
            customKeypad: true,
        });

        expect(container).toMatchSnapshot("mobile render");
    });

    it("can be focused", async () => {
        const {renderer} = renderQuestion(question1);

        // Act
        const gotFocus = await act(() => renderer.focus());

        // Assert
        expect(gotFocus).toBe(true);
        expect(screen.getByRole("textbox", {hidden: true})).toHaveFocus();
    });

    it("can be blurred", () => {
        const {renderer} = renderQuestion(question1);

        // Act
        const input = screen.getByRole("textbox", {hidden: true});
        act(() => input.focus());
        act(() => renderer.blur());

        // Assert
        // eslint-disable-next-line testing-library/no-node-access
        expect(document.activeElement).not.toBe(
            screen.getByRole("textbox", {hidden: true}),
        );
    });
});

describe("unionAnswerForms utility function", () => {
    beforeEach(() => {
        jest.spyOn(Dependencies, "getDependencies").mockReturnValue(
            testDependencies,
        );
    });

    it("removes duplicates", () => {
        // arrange
        const forms = [
            [
                {
                    simplify: "required" as const,
                    name: "integer",
                } as const,
            ],
            [
                {
                    simplify: "required" as const,
                    name: "integer",
                } as const,
            ],
        ];

        // act
        const result = unionAnswerForms(forms);

        // assert
        expect(result).toHaveLength(1);
    });
});

describe("maybeParsePercentInput utility function", () => {
    beforeEach(() => {
        jest.spyOn(Dependencies, "getDependencies").mockReturnValue(
            testDependencies,
        );
    });

    it("returns a percent if the input is a percent", () => {
        const result = maybeParsePercentInput("3%", false);

        expect(result).toBe(3);
    });

    it("returns the input if its not valid", () => {
        const result = maybeParsePercentInput("asdasd%", false);

        expect(result).toBe("asdasd%");
    });

    it("normalizes values", () => {
        const result = maybeParsePercentInput("20%", true);

        expect(result).toBe(0.2);
    });
});
