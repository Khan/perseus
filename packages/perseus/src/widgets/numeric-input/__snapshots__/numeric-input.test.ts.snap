// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Numeric input widget styles differently on mobile: mobile render 1`] = `
<div>
  <div
    class="perseus-renderer perseus-renderer-responsive"
  >
    <div
      class="paragraph"
      data-perseus-paragraph-index="0"
    >
      <div
        class="paragraph"
      >
        <span
          style="white-space: nowrap;"
        >
          <span />
          <span
            class="mock-TeX"
          >
            5008 \\div 4 =
          </span>
          <span />
        </span>
         
        <div
          class="perseus-widget-container widget-nohighlight widget-inline-block"
        >
          <div>
            <div
              aria-label="Math input box Tap with one or two fingers to open keyboard"
              class="initial_4qg14c-o_O-input_1n1c032"
              role="textbox"
            >
              <div
                class="keypad-input"
                tabindex="0"
              >
                <div
                  class="mq-editable-field mq-math-mode"
                  style="background-color: white; min-height: 44px; min-width: 64px; max-width: 128px; box-sizing: border-box; position: relative; border-style: solid; border-color: rgba(33,36,44,0.50); border-radius: 4px; color: rgb(33, 36, 44); border-width: 1px;"
                >
                  <span
                    class="mq-textarea"
                  >
                    <span
                      aria-label="Math Input:"
                    />
                  </span>
                  <span
                    aria-hidden="true"
                    class="mq-root-block mq-empty"
                    style="padding: 10px 13px 8px 13px; font-size: 18pt;"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
         
      </div>
    </div>
  </div>
</div>
`;

exports[`numeric-input widget Should render predictably: after interaction 1`] = `
<div>
  <div
    class="perseus-renderer perseus-renderer-responsive"
  >
    <div
      class="paragraph"
      data-perseus-paragraph-index="0"
    >
      <div
        class="paragraph"
      >
        <span
          style="white-space: nowrap;"
        >
          <span />
          <span
            class="mock-TeX"
          >
            5008 \\div 4 =
          </span>
          <span />
        </span>
         
        <div
          class="perseus-widget-container widget-nohighlight widget-inline-block"
        >
          <input
            aria-disabled="false"
            aria-invalid="false"
            aria-required="false"
            autocapitalize="off"
            autocomplete="off"
            autocorrect="off"
            class="input_1ck1z8k-o_O-LabelMedium_1rew30o-o_O-default_1tc7wjm-o_O-defaultFocus_lrnqlz-o_O-inputWithExamples_ylhnsi"
            id=":r6:"
            tabindex="0"
            type="text"
            value="1252"
          />
          <span
            id="aria-for-:r6:"
            style="display: none;"
          />
        </div>
         
      </div>
    </div>
  </div>
</div>
`;

exports[`numeric-input widget Should render predictably: first render 1`] = `
<div>
  <div
    class="perseus-renderer perseus-renderer-responsive"
  >
    <div
      class="paragraph"
      data-perseus-paragraph-index="0"
    >
      <div
        class="paragraph"
      >
        <span
          style="white-space: nowrap;"
        >
          <span />
          <span
            class="mock-TeX"
          >
            5008 \\div 4 =
          </span>
          <span />
        </span>
         
        <div
          class="perseus-widget-container widget-nohighlight widget-inline-block"
        >
          <input
            aria-disabled="false"
            aria-invalid="false"
            aria-required="false"
            autocapitalize="off"
            autocomplete="off"
            autocorrect="off"
            class="input_1ck1z8k-o_O-LabelMedium_1rew30o-o_O-default_1tc7wjm-o_O-defaultFocus_lrnqlz-o_O-inputWithExamples_1y6ajxo"
            id=":r6:"
            tabindex="0"
            type="text"
            value=""
          />
          <span
            id="aria-for-:r6:"
            style="display: none;"
          />
        </div>
         
      </div>
    </div>
  </div>
</div>
`;

exports[`numeric-input widget Should render tooltip as list when multiple format options are given: render with format list tooltip 1`] = `
<div>
  <div
    class="perseus-renderer perseus-renderer-responsive"
  >
    <div
      class="paragraph"
      data-perseus-paragraph-index="0"
    >
      <div
        class="paragraph"
      >
        <span
          style="white-space: nowrap;"
        >
          <span />
          <span
            class="mock-TeX"
          >
            5008 \\div 4 =
          </span>
          <span />
        </span>
         
        <div
          class="perseus-widget-container widget-nohighlight widget-inline-block"
        >
          <input
            aria-describedby="aria-for-:rf:"
            aria-disabled="false"
            aria-invalid="false"
            aria-required="false"
            autocapitalize="off"
            autocomplete="off"
            autocorrect="off"
            class="input_1ck1z8k-o_O-LabelMedium_1rew30o-o_O-default_1tc7wjm-o_O-defaultFocus_lrnqlz-o_O-inputWithExamples_1y6ajxo"
            id=":rf:"
            tabindex="0"
            type="text"
            value=""
          />
          <span
            id="aria-for-:rf:"
            style="display: none;"
          >
            Your answer should be 
                   a simplified proper fraction, like 3/5, or
a simplified improper fraction, like 7/4, or
a mixed number, like 1 and 3/4
          </span>
        </div>
         
      </div>
    </div>
  </div>
</div>
`;

exports[`numeric-input widget Should render tooltip using only correct answer formats: render tooltip only with correct answers 1`] = `
<div>
  <div
    class="perseus-renderer perseus-renderer-responsive"
  >
    <div
      class="paragraph"
      data-perseus-paragraph-index="0"
    >
      <div
        class="paragraph"
      >
        <span
          style="white-space: nowrap;"
        >
          <span />
          <span
            class="mock-TeX"
          >
            5008 \\div 4 =
          </span>
          <span />
        </span>
         
        <div
          class="perseus-widget-container widget-nohighlight widget-inline-block"
        >
          <input
            aria-describedby="aria-for-:ri:"
            aria-disabled="false"
            aria-invalid="false"
            aria-label="What's the answer?"
            aria-required="false"
            autocapitalize="off"
            autocomplete="off"
            autocorrect="off"
            class="input_1ck1z8k-o_O-LabelMedium_1rew30o-o_O-default_1tc7wjm-o_O-defaultFocus_lrnqlz-o_O-inputWithExamples_1y6ajxo"
            id=":ri:"
            tabindex="0"
            type="text"
            value=""
          />
          <span
            id="aria-for-:ri:"
            style="display: none;"
          >
            Your answer should be 
                   a simplified proper fraction, like 3/5
          </span>
        </div>
         
      </div>
    </div>
  </div>
</div>
`;

exports[`numeric-input widget Should render tooltip when format option is given: render with format tooltip 1`] = `
<div>
  <div
    class="perseus-renderer perseus-renderer-responsive"
  >
    <div
      class="paragraph"
      data-perseus-paragraph-index="0"
    >
      <div
        class="paragraph"
      >
        <span
          style="white-space: nowrap;"
        >
          <span />
          <span
            class="mock-TeX"
          >
            5008 \\div 4 =
          </span>
          <span />
        </span>
         
        <div
          class="perseus-widget-container widget-nohighlight widget-inline-block"
        >
          <input
            aria-describedby="aria-for-:r9:"
            aria-disabled="false"
            aria-invalid="false"
            aria-required="false"
            autocapitalize="off"
            autocomplete="off"
            autocorrect="off"
            class="input_1ck1z8k-o_O-LabelMedium_1rew30o-o_O-default_1tc7wjm-o_O-defaultFocus_lrnqlz-o_O-inputWithExamples_1y6ajxo"
            id=":r9:"
            tabindex="0"
            type="text"
            value=""
          />
          <span
            id="aria-for-:r9:"
            style="display: none;"
          >
            Your answer should be 
                   a simplified proper fraction, like 3/5
          </span>
        </div>
         
      </div>
    </div>
  </div>
</div>
`;
